torch/csrc/dynamo/eval_frame.c

```diff
@@ -745,17 +798,17 @@ static PyObject* _custom_eval_frame(
     return NULL;
   } else if (result != Py_None) {
     DEBUG_TRACE("create cache %s", get_frame_name(frame));
-    extra = create_cache_entry(extra, result);
+    CacheEntry* new_cache_entry = create_cache_entry(cache_entry, result);
     Py_DECREF(result);
-    set_cache_entry(frame->f_code, extra);
+    create_and_set_extra_state(frame->f_code, new_cache_entry, frame_state);
     // Re-enable custom behavior
     eval_frame_callback_set(callback);
-    return eval_custom_code(tstate, frame, extra->code, throw_flag);
+    return eval_custom_code(tstate, frame, new_cache_entry->code, throw_flag);
   } else {
     DEBUG_TRACE("create skip %s", get_frame_name(frame));
     Py_DECREF(result);
-    destroy_cache_entry(extra);
-    set_cache_entry(frame->f_code, SKIP_CODE);
+    destroy_extra_state(frame->f_code);
+    set_extra_state(frame->f_code, SKIP_CODE);
```

Comment:
The destroy above shouldn't be necessary, right? I would expect `set_extra_state` to care of deallocating the previous entry.

Reply:
`set_extra_state` is really simple, it just hides a private API - `_PyCode_SetExtra`

```
inline static void set_extra_state(PyCodeObject* code, ExtraState* extra_state) {
  // Sets the extra state on the extra scrach space of the code object.
  _PyCode_SetExtra((PyObject*)code, extra_index, extra_state);
}
```

I don't want set_extra_state` to do deallocation. One could just alter the items in the linked list and then set the extra_state. In that case, we would not want to deallocate, we would just want to override. So, having different functions handle the deallocation and setting the extra field seems ok to me.

Reply:
OK, then you should document that set_extra_state does NOT deallocate the previous entry and it's the responsibility of the caller to dispose of the previous entry.

Reply:
I went back and made `set_extra_state` to deallocate the prev entry.